ls1intum/Artemis
# `Programming exercises`: Check if a build plan already exists before creating it when using Jenkins

### Checklist
#### General


#### Server
#### Changes affecting Programming Exercises

### Motivation and Context


During our lectures, we found out that it may be possible that a  Job in Jenkins is not linked to artemis.
Therefore, Artemis crashes with an internal server error.

### Description
I've added a simple check whether a job already exists. If so, no new job will be created.



### Steps for Testing

Prerequisites:
- Access to the build server
- Instructor Access
- 2 Students (1 is enough, but 2 are easier)

1. Create a programming task
2. Manually create a valid Job in Jenkins that matches the Name & Folder of a student's account (called student1 in the following) (you can also start a participation of another student and copy the settings to retrieve a valid one)
3. Try to start the task as the student1 .. you should get an Internal Server Error before this patch. Now the task should be started and builds should be successful.

### Review Progress





#### Code Review
#### Manual Tests

### Screenshots of original Issue
![image](https://user-images.githubusercontent.com/22198798/141468828-bda4c6c2-72f0-449a-b076-1a6b1e73cd0a.png)
![image](https://user-images.githubusercontent.com/22198798/141468912-86188805-1b05-4087-b3dd-7a569c334d30.png)

```
de.tum.in.www1.artemis.exception.JenkinsException: status code: 400, reason phrase: Bad Request
        at de.tum.in.www1.artemis.service.connectors.jenkins.jobs.JenkinsJobService.createJobInFolder(JenkinsJobService.java:146)
        at de.tum.in.www1.artemis.service.connectors.jenkins.JenkinsBuildPlanService.copyBuildPlan(JenkinsBuildPlanService.java:230)
        at de.tum.in.www1.artemis.service.connectors.jenkins.JenkinsService.copyBuildPlan(JenkinsService.java:91)
        at de.tum.in.www1.artemis.service.ParticipationService.copyBuildPlan(ParticipationService.java:381)
        at de.tum.in.www1.artemis.service.ParticipationService.startProgrammingExercise(ParticipationService.java:171)
        at de.tum.in.www1.artemis.service.ParticipationService.startExercise(ParticipationService.java:131)
        at de.tum.in.www1.artemis.service.ParticipationService$$FastClassBySpringCGLIB$$f26b3f60.invoke(<generated>)
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
```
